src/meson.build

```diff
@@ -0,0 +1,425 @@
+  'multicast_device.c',
+  'net.c',
+  'net_packet.c',
+  'net_setup.c',
+  'net_socket.c',
+  'node.c',
+  'process.c',
+  'protocol.c',
+  'protocol_auth.c',
+  'protocol_edge.c',
+  'protocol_key.c',
+  'protocol_misc.c',
+  'protocol_subnet.c',
+  'raw_socket_device.c',
+  'route.c',
+  'subnet.c',
+  'tincd.c',
+  common_src,
+]
+
+sptps_test_src = [
+  'ed25519/ecdh.c',
+  'ed25519/ecdsa.c',
+  'logger.c',
+  'sptps.c',
+  'sptps_test.c',
+  'utils.c',
+  'xoshiro.c',
+  chacha_poly_src,
+  ed25519_src,
+]
+
+sptps_keypair_src = [
+  'ed25519/ecdsagen.c',
+  'sptps_keypair.c',
+  'utils.c',
+  ed25519_src,
+]
+
+sptps_speed_src = [
+  'ed25519/ecdh.c',
+  'ed25519/ecdsa.c',
+  'ed25519/ecdsagen.c',
+  'logger.c',
+  'sptps.c',
+  'sptps_speed.c',
+  'utils.c',
+  chacha_poly_src,
+  ed25519_src,
+]
+
+tincd_cc_flags = [cc_flags]
+
+tinc_dep = []
+tincd_dep = [
+  cc.find_library('m', required: false),
+]
+sptps_speed_dep = []
+sptps_test_dep = []
+sptps_keypair_dep = []
+splice_dep = []
+
+if os_name == 'linux'
+  subdir('linux')
+elif os_name.endswith('bsd') or os_name in ['dragonfly', 'darwin']
+  subdir('bsd')
+elif os_name == 'sunos'
+  subdir('solaris')
+elif os_name == 'windows'
+  subdir('mingw')
+endif
+
+foreach h : check_headers
+  if cc.has_header(h)
+    cdata.set('HAVE_' + h.to_upper().underscorify(),
+              1,
+              description: '#include <' + h + '>')
+  endif
+endforeach
+
+confdata = configuration_data()
+confdata.merge_from(cdata)
+configure_file(output: 'meson_config.h', configuration: confdata)
+
+src_dir = meson.current_source_dir()
+have_h = src_dir / 'have.h'
+
+have_prefix = '''
+#include "@0@/src/meson_config.h"
+#include "@1@"
+'''.format(build_root, have_h)
+
+foreach f : check_functions
+  if f == 'fork' and os_name == 'windows'
+    message('MinGW does not have correct definition for fork()')
+  else
+    if cc.has_function(f, prefix: have_prefix, args: cc_defs)
+      cdata.set('HAVE_' + f.to_upper(),
+                1,
+                description: 'function ' + f)
+    endif
+  endif
+endforeach
+
+if cc.has_function('res_init', prefix: '''
+  #include <netinet/in.h>
+  #include <resolv.h>
+''', args: cc_defs)
+  cdata.set('HAVE_DECL_RES_INIT', 1)
+endif
+
+foreach type : check_types
+  if cc.has_type(type, prefix: have_prefix, args: cc_defs)
+    name = 'HAVE_' + type.to_upper().underscorify()
+    cdata.set(name, 1, description: type)
+  endif
+endforeach
+
+if not cdata.has('HAVE_GETOPT_H') or not cc.has_function('getopt_long', prefix: have_prefix, args: cc_defs)
+  src = ['getopt.c', 'getopt1.c']
+  tinc_src += src
+  tincd_src += src
+  sptps_test_src += src
+  sptps_keypair_src += src
+endif
+
+# No pkg-config files on MinGW
+opt_miniupnpc = get_option('miniupnpc')
+dep_miniupnpc = dependency('miniupnpc', required: opt_miniupnpc, static: win_static)
+if not dep_miniupnpc.found()
+  dep_miniupnpc = cc.find_library('miniupnpc', required: opt_miniupnpc, static: win_static)
+endif
```

Comment:
If you pass `-D miniupnpc=enabled`, then if `dependency()` doesn't find the MiniUPnP library, it will immediately stop with an error, it will not try the `cc.find_library()` method. Also, if you pass `-D miniupnpc=disabled`, it will do both checks. So perhaps it should be rewritten to:

```
if opt_miniupnpc.allowed():
    dep_miniupnpc = dependency('miniupnpc', required: false, static: win_static)
    if not dep.miniupnpc.found()
        dep_miniupnpc = cc.find_library('miniupnpc', required: opt_miniupnpc, static: win_static)
    endif
    if dep.miniupnpc.found()
         # Add dependency to tincd_dep
         ...
    endif
endif
```